### ISSUE TITLE
Run phrase extraction on on the text from PDFs

### ISSUE THREAD

**Issue by ceteri** [Coleridge-Initiative/rclc#20]
For the next stage to add to this pipeline, run phrase extraction from the JSON files that result from #10.

See <https://pypi.org/project/pytextrank/> for a pipeline based on [spaCy](https://spacy.io/)
with example code at <https://github.com/DerwenAI/pytextrank#usage>

The objective is to determine the top-ranked phrases for each section of a research paper. If possible, also collect the title of the section -- then save results out to JSON.

**Comment by JasonZhangzy1757** [comment]
#### Hi @ceteri 

Here is [a sample result](https://github.com/JasonZhangzy1757/rclc/blob/master/%23%2020%20phrase%20extraction/output_sample.json) of phrase extraction for each section with its title and [the code](https://github.com/JasonZhangzy1757/rclc/blob/master/%23%2020%20phrase%20extraction/Phrase%20Extraction%20.ipynb) is attached under the same directory. I haven't made it well-organized yet, but it generally used the code from the  `parsr_output_interpreter.py` and `explain_algo.ipynb` and updated a little for our needs. The [article](https://github.com/JasonZhangzy1757/rclc/blob/master/%23%2020%20phrase%20extraction/PE_sample.pdf) and [JSON](https://github.com/JasonZhangzy1757/rclc/blob/master/%23%2020%20phrase%20extraction/PE_sample.pdf.json) from Parsr are also attached for reference. 

Please check if I understand you request correctly :)

**A little concern:**  We might have a potential problem, because the titles are extracted where the `type` of element is `heading` in the JSON output, and the sections are divided following the same rule. But I noticed there are missing or wrong `heading` labels, and in some research paper this misclassification could be quite often.  Thus this might affects the performance of future section classification tasks. Yet I assume this is an intrinsic error due to the limit of the parser…  so I think at least I should let you know.

**Comment by ceteri** [comment]
Thanks @JasonZhangzy1757  -

For this work, use the `spaCy` pipeline rather than reimplementing the `pytextrank` library. Here's some example code: <https://github.com/DerwenAI/pytextrank/blob/master/example.py>

**Comment by ceteri** [comment]
Also, that JSON file looks very different than the previous sample output from Parsr -- is that a different setting?  It appears to be the translated events for rendering PDF, not the extracted text?

**Comment by JasonZhangzy1757** [comment]
#### Hi @ceteri
:( Oh ... Seems I thought it too simply. Sorry about that and I'll keep working on that.

As for the JSON, I think they should be the same because they are generated using the same code with the same client and config. And I think the only difference between the two is that the `metadata` of this one is empty, so it looks a bit different. If you search term `page` in the previous sample output and compare, they are actually in the same format. Or are there other differences I didn't notice?

Also I'll look into why this one doesn't have `metadata`.

**Comment by JasonZhangzy1757** [comment]
#### Hi @ceteri 

I've now used the spaCy pipeline for the code. Please check if [this](https://github.com/JasonZhangzy1757/rclc/blob/master/%23%2020%20phrase%20extraction/Phrase%20Extraction%20-%20spaCy%20pipeline.py) is what you want. So basically I've added these code:
```javascript
# Add PyTextRank into the spaCy pipeline
tr = pytextrank.TextRank(logger=None)
nlp.add_pipe(tr.PipelineComponent, name="textrank", last=True)
```

Also, should I make the script walking through all the JSON output in the directory so that it could add to the workflow later as well?

**Comment by ceteri** [comment]
Nice work on that!  The results are really interesting, and the top-ranked phrases seemed to fit with the main "keywords" identified for the paper by its authors.

Yes, for next steps walk through all of the JSON output in the directory.  We'll use the results of this part internally -- not pushing it up to the S3 bucket.

I have a hunch that output from `Parsr` will tend to have hard-breaks for hyphenations. That's probably more a matter of how these papers get formatted into two columns by LaTex, etc. In any case, we may be able to use a filter to convert the hyphen/space: `a geo- privacy protection` => `a geo-privacy protection` 

Also, your functions `FindObject()` and `GetText()` would be good to move into a library, which this script uses. That could help make it simpler for other people to integrate with the JSON results from `Parsr` ?

**Comment by JasonZhangzy1757** [comment]
No Problem! I'll work on that.

**Comment by JasonZhangzy1757** [comment]
#### Hi @ceteri 

 - [Here](https://github.com/Coleridge-Initiative/rclc/blob/zz2498/Parsr/bin/parsed_json_interpreter.py) is the library I created. 
 -  And [here](https://github.com/Coleridge-Initiative/rclc/blob/zz2498/Parsr/bin/phrase_extraction_pipeline.py) is  the code walking through all the JSON output by Parsr and outputting TextRank JSON using the library I created based on spaCy pipeline. 

So now after Parsr converts PDF to text and JSON, we only need to run this line, and we could get TextRank JSON for each section of every article in the directory.
```
python bin/phrase_extraction_pipeline.py
```

Some other points I think worthing mentioning:
 - It seems most of the files created by Parsr has metadata, and it won't affect the results of TextRank.
 - Sometimes the `heading`  is misclassified by Parsr, but for later classification we could focus only on typical titles, like Introduction, Method and so on. I think this could reduce the negative impact of misclassification.
 - I used `en_core_web_sm` here for efficiency, but we could use medium or large language model later in the workflow.
 - This time I put the scripts in the branch :)

**Comment by ceteri** [comment]
Great, thanks!  Using `en_core_web_sm` is fine for this application -- the larger language models appear to be more for other parts of spaCy.

I'll answer over on the PR #22

**Comment by ceteri** [comment]
implemented by PR #22
